plugins/BEdita/API/src/Middleware/AnalyticsMiddleware.php

```diff
@@ -0,0 +1,153 @@
+<?php
+/**
+ * BEdita, API-first content management framework
+ * Copyright 2017 ChannelWeb Srl, Chialab Srl
+ *
+ * This file is part of BEdita: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * See LICENSE.LGPL or <http://gnu.org/licenses/lgpl-3.0.html> for more details.
+ */
+namespace BEdita\API\Middleware;
+
+use BEdita\Core\State\CurrentApplication;
+use BEdita\Core\Utility\LoggedUser;
+use Cake\Event\EventDispatcherTrait;
+use Cake\Log\Log;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+
+/**
+ * Middleware to trace analytics data
+ *
+ * @since 4.0.0
+ */
+class AnalyticsMiddleware
+{
+    use EventDispatcherTrait;
+
+    /**
+     * Request start time
+     *
+     * @var float
+     */
+    protected $startTime = null;
+
+    /**
+     * Analytics data
+     *
+     * @var array
+     */
+    protected $data = [];
+
+    /**
+     * Configure analytics logger if not configured yet
+     *
+     */
+    public function __construct()
+    {
+        $this->startTime = microtime(true);
+        if (defined('TIME_START')) {
+            $this->startTime = TIME_START;
+        }
+
+        if (!in_array('analytics', Log::configured())) {
+            Log::config('analytics', [
+                'className' => 'File',
+                'path' => LOGS,
+                'scopes' => ['analytics'],
+                'file' => 'analytics',
+            ]);
+        }
+    }
+
+    /**
+     * Getter for $startTime
+     *
+     * @return float
+     * @codeCoverageIgnore
+     */
+    public function getStartTime()
+    {
+        return $this->startTime;
+    }
+
+    /**
+     * Getter for $data
+     *
+     * @return array
+     * @codeCoverageIgnore
+     */
+    public function getData()
+    {
+        return $this->data;
+    }
+
+    /**
+     * Get analytics custom data from event listeners
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request.
+     * @param \Psr\Http\Message\ResponseInterface $response The response.
+     * @return array
+     */
+    protected function readCustomData(ServerRequestInterface $request, ResponseInterface $response)
+    {
+        $event = $this->dispatchEvent('Analytics.custom', [$request, $response]);
+        if (empty($event->result)) {
+            return [];
```

Comment:
Just a side note: empty arrays are always encoded by `json_encode` as `[]`. For some parses this can be troublesome if they expect an object. If we want empty value to be encoded as `{}` we might return `new \stdClass()` instead of the empty array, to be sure that `"x"` is always an object.